### PR TITLE
OM: Validate utf8 in labelvalues

### PR DESCRIPTION
**Pull request by leecalcote**
Working with @girishranganathan to knock out a TODO for OpenMetrics compliant client.

### CODE REVIEW

File: prometheus_client/openmetrics/parser.py [hunk 1]
```diff
@@ -136,6 +140,8 @@ def _parse_labels(it, text):
                 raise ValueError("Invalid line: " + text)
     return labels
 
+def _validate_utf8(char):
+    return ord(char) < 65536
```
Comment (brian-brazil) [inline prometheus_client/openmetrics/parser.py]: Why is this a test for a utf-8 character?

I think you were going the right direction with the encode.

Comment (leecalcote) [reply prometheus_client/openmetrics/parser.py]: @brian-brazil ok. Starting on line 109, how's this instead?

`utf8_str = ''.join(labelvalue)
                try:
                    utf8_str.encode('utf')
                except:
                    raise ValueError("Invalid line: " + text`

Comment (brian-brazil) [reply prometheus_client/openmetrics/parser.py]: That sounds about right.

File: prometheus_client/openmetrics/parser.py [hunk 2]
```diff
@@ -106,11 +106,16 @@ def _parse_labels(it, text):
             elif char == '"':
                 if not core._METRIC_LABEL_NAME_RE.match(''.join(labelname)):
                     raise ValueError("Invalid line: " + text)
-                labels[''.join(labelname)] = ''.join(labelvalue)
+                utf8_str = ''.join(labelvalue)
+                try:
+                    utf8_str.encode('utf')
```
Comment (brian-brazil) [inline prometheus_client/openmetrics/parser.py]: I believe it's `utf-8`

**Comment by brian-brazil** [comment]
That looks right, the test you added is failing though.

File: prometheus_client/openmetrics/parser.py [hunk 3]
```diff
@@ -106,11 +106,16 @@ def _parse_labels(it, text):
             elif char == '"':
                 if not core._METRIC_LABEL_NAME_RE.match(''.join(labelname)):
                     raise ValueError("Invalid line: " + text)
-                labels[''.join(labelname)] = ''.join(labelvalue)
+                utf8_str = ''.join(labelvalue)
+                try:
+                    utf8_str.encode('utf-8')
+                except:
```
Comment (brian-brazil) [inline prometheus_client/openmetrics/parser.py]: Catch the specific exception type that you're looking for

Comment (leecalcote) [reply prometheus_client/openmetrics/parser.py]: @brian-brazil we're scratching our heads on finding a set of characters as a test case...

Comment (leecalcote) [reply prometheus_client/openmetrics/parser.py]: Added UnicodeEncodeError exception type.

File: prometheus_client/openmetrics/parser.py [hunk 4]
```diff
@@ -111,11 +112,19 @@ def _parse_labels(it, text):
             elif char == '"':
                 if not core._METRIC_LABEL_NAME_RE.match(''.join(labelname)):
                     raise ValueError("Invalid line: " + text)
-                labels[''.join(labelname)] = ''.join(labelvalue)
+                utf8_str = ''.join(labelvalue)
+                try:
+                    if sys.version_info >= (3,):
+                        utf8_str.encode('utf-8')
+                    else:
+                        utf8_str.decode('utf-8')
```
Comment (brian-brazil) [inline prometheus_client/openmetrics/parser.py]: This doesn't make sense to me, it should be an encode in both cases.

Comment (brian-brazil) [reply prometheus_client/openmetrics/parser.py]: Maybe you need a different test for Python 2

File: .gitignore [hunk 5]
```diff
@@ -7,4 +7,5 @@ dist
 .coverage
 .tox
 .*cache
-htmlcov
\ No newline at end of file
+htmlcov
+.vscode
```
Comment (brian-brazil) [inline .gitignore]: This should go in your global gitignore

File: tests/openmetrics/test_parser.py [hunk 6]
```diff
@@ -247,6 +247,22 @@ def test_labels_with_curly_braces(self):
         metric_family.add_metric(["bar", "b{a}z"], 1)
         self.assertEqual([metric_family], list(families))
 
+    def test_labels_with_invalid_utf8_values(self):
+        try:
+            if sys.version_info < (2, 7):
```
Comment (brian-brazil) [inline tests/openmetrics/test_parser.py]: I think you want to go back to >= 3

Not sure why this is failing on 2.6

File: tests/openmetrics/test_parser.py [hunk 7]
```diff
@@ -252,7 +252,7 @@ def test_labels_with_invalid_utf8_values(self):
             if sys.version_info >= (3,):
                 inj = u'\uD802'
             else:
-                inj = '\xf8\xa1\xa1\xa1\xa1'
+                inj = '\xfc'
```
Comment (brian-brazil) [inline tests/openmetrics/test_parser.py]: \xff should do it

Comment (girishranganathan) [reply tests/openmetrics/test_parser.py]: so using this or any in this way dint help but using them as a binary string seemed to have done the trick

**Comment by brian-brazil** [comment]
Thanks!

**Comment by leecalcote** [comment]
Excellent. :)

**Pull request merged**